controllers/nstemplateset/nstemplateset_controller_test.go

```diff
@@ -968,6 +968,63 @@ func TestDeleteNSTemplateSet(t *testing.T) {
 		AssertThatNSTemplateSet(t, namespaceName, username, r.Client).
 			DoesNotHaveFinalizer() // finalizer was not added and nothing else was done
 	})
+
+	t.Run("NSTemplateSet not deleted until namespace is deleted", func(t *testing.T) {
+		// given an NSTemplateSet resource and 1 active user namespaces ("dev")
+		nsTmplSet := newNSTmplSet(namespaceName, username, "advanced", withNamespaces("abcde11", "dev"), withDeletionTs(), withClusterResources("abcde11"))
+		devNS := newNamespace("advanced", username, "dev", withTemplateRefUsingRevision("abcde11"))
+
+		r, fakeClient := prepareController(t, nsTmplSet, devNS)
+		req := newReconcileRequest(namespaceName, username)
+
+		// only add deletion timestamp, but not delete
+		fakeClient.MockDelete = func(ctx context.Context, obj client.Object, opts ...client.DeleteOption) error {
+			if obj, ok := obj.(*corev1.Namespace); ok {
+				deletionTs := metav1.Now()
+				obj.DeletionTimestamp = &deletionTs
+				if err:= r.Client.Update(context.TODO(), obj); err !=nil {
+					return err
+				}
+			}
+			return nil
+		}
+		// first reconcile, deletion is triggered
+		_, err := r.Reconcile(context.TODO(), req)
+		require.NoError(t, err)
+		//then second reconcile to check if namespace has actually been deleted
+		_, err = r.Reconcile(context.TODO(), req)
+		require.NoError(t, err)
+
+		// get the first namespace and check that it is not deleted
+		firstNSName := fmt.Sprintf("%s-dev", username)
+		AssertThatNamespace(t, firstNSName, r.Client).DoesExist()
+		// get the NSTemplateSet resource again, check it is not deleted and its status
+		AssertThatNSTemplateSet(t, namespaceName, username, r.Client).
+			HasFinalizer().
+			HasConditions(UnableToTerminate("user namespace johnsmith-dev deletion was triggered but is not complete yet, something could be blocking ns deletion"))
+
+		// remove deletion timestamp, actual delete is now triggered
+		userNS := &corev1.Namespace{}
+		err = r.Client.Get(context.TODO(), types.NamespacedName{Name: firstNSName}, userNS)
+		require.NoError(t, err)
+		userNS.DeletionTimestamp = nil
+		err = r.Client.Update(context.TODO(), userNS)
+		require.NoError(t, err)
```

Comment:
What you are doing is actually making the namespace active again and the controller delete it again but with a regular Delete() in the fake client. This is not what we want. Instead of this we need to actually delete the resource from the fake client here, in the test. In this case you would simulate the actual deletion by kube.